Archive/Draft/RFCxxxx-PS7-Bootstrapper.md

```diff
@@ -0,0 +1,256 @@
+---
+RFC: RFCnnnn
+Author: Steve Lee
+Status: Draft
+SupercededBy: n/a
+Version: 0.1
+Area: Install
+Comments Due: Dec 31st, 2021
+Plan to implement: Yes
+---
+
+# Bootstrapper for PS7 in Windows
+
+## Motivation
+
+    As an admin,
+    I can use PS7 to provision a new Windows machine,
+    so that I can leverage new capabilities in PS7.
+
+    As a user,
+    I can use PS7 on a new Windows machine,
+    so that I can leverage new capabilities in PS7.
+
+Explicitly installing PS7 is a blocker for some customers not only due to discoverability, but also some customers have policies that
+don't allow for installing software onto Windows after initial provisioning.
+
+## Shipping in Windows
+
+There are several issues that prevent simply shipping PS7 in Windows:
+
+- Size
+- Support lifecycle differences
+
+### Size constraints
+
+A complete installation of PowerShell 7 currently is around 200 MB (unpacked).
+This is due to including everything (modules and .NET assemblies) to provide parity with what Windows PowerShell 5.1 with .NET Framework provides.
+There is currently no plan by Windows or the .NET team to ship newer .NET in Windows, so PS7 will need to continue to ship .NET with PS7.
+
+### Support lifecycle
+
+Windows has a 5+5 year support lifecycle.
+.NET supports up to 3 years for Long Term Support (LTS) versions.
+So even if size was not an issue, we cannot check in a version of PS7 with a .NET that is not supported while Windows itself is supported.
+
+### Solution
+
+To solve both of these issues, the proposal is to ship a bootstrapper in Windows that would be supported within the Windows support lifecycle.
+The boostrapper will update PS7 as needed to ensure a supported version of .NET is being used.
+Details of this bootstrapper follow.
+
+## Specification
+
+### pwsh.exe bootstrapper
+
+We would check-in a `pwsh.exe` into Windows 11+ (under `System32`) written in [native code](#native-code) that will:
+
+- Detect if PS7 is already installed
+  - If installed, pass-through all arguments to the new detached `pwsh` child process
+    - This means that if PS7 is already in the `PATH`, then that version (whether installed via MSI, MSIX, or other means)
+      needs to be updated outside of the bootstrapper
+  - Otherwise, [install](#installer) PS7 and then pass-through all arguments to the new detached `pwsh` child process
```

Comment:
Feels like this would essentially circumvent software installation policies established by the enterprise and - in a worst case scenario - might become yet another hole enterprise will now need to be closing.

Reply:
how would one control aspects of such installation? E.g. things like path, whether it is single/multi user install, logging and verbosity

Reply:
There is another issue that comes to mind: what about software that first checks for the *presence* of pwsh.exe, and then calls "pwsh.exe --version" to detect the *installed* PowerShell version? Not only the shim would fool most software into thinking it's already installed, but the "pwsh.exe --version" version call would trigger the actual installation.

Reply:
Yep, this is the same problem the Python for Windows folks created. They have a python.exe alias that build scripts detected as "Python is installed". Terrible dev experience.

Reply:
In this case, PS7 should just be considered a Windows feature-on-demand (even though it doesn't use that mechanism specifically because a literal Feature-on-Demand requires the 10 year support lifecycle) and would install via Windows Update channel (this is in the RFC).  So for all intents and purposes, pwsh.exe is in-box regardless if it needs to bootstrap itself initially.  `pwsh -version` would work as though it was installed.